File: apps/frontend/fonte/services/api/documentos-rh.js
import { requisitar, requisitarArquivo } from './core.js';

export async function listarDocumentosRh({
  idPastaPai = null,
  tipo = '',
  categoriaExtensao = '',
  busca = '',
  criadoPor = '',
  dataCriacaoDe = '',
  dataCriacaoAte = '',
  dataModificacaoDe = '',
  dataModificacaoAte = '',
} = {}) {
  const params = new URLSearchParams();
  if (idPastaPai !== null && idPastaPai !== undefined && idPastaPai !== '') {
    params.set('id_pasta_pai', idPastaPai);
  }
  if (tipo) params.set('tipo', tipo);
  if (categoriaExtensao) params.set('categoria_extensao', categoriaExtensao);
  if (busca) params.set('busca', busca);
  if (criadoPor) params.set('criado_por', criadoPor);
  if (dataCriacaoDe) params.set('data_criacao_de', dataCriacaoDe);
  if (dataCriacaoAte) params.set('data_criacao_ate', dataCriacaoAte);
  if (dataModificacaoDe) params.set('data_modificacao_de', dataModificacaoDe);
  if (dataModificacaoAte) params.set('data_modificacao_ate', dataModificacaoAte);

  const sufixo = params.toString() ? `?${params.toString()}` : '';
  return requisitar(`/documentos-rh${sufixo}`, { method: 'GET' });
}

export async function criarPastaDocumentoRh({ nome, idPastaPai = null }) {
  return requisitar('/documentos-rh/pastas', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ nome, id_pasta_pai: idPastaPai }),
  });
}

export async function uploadArquivoDocumentoRh(arquivo, idPastaPai = null) {
  const formData = new FormData();
  formData.append('arquivo', arquivo);

  const params = new URLSearchParams();
  if (idPastaPai !== null && idPastaPai !== undefined && idPastaPai !== '') {
    params.set('id_pasta_pai', idPastaPai);
  }
  const sufixo = params.toString() ? `?${params.toString()}` : '';

  return requisitar(`/documentos-rh/upload${sufixo}`, {
    method: 'POST',
    body: formData,
  });
}

export async function obterConteudoDocumentoRh(idDocumento) {
  return requisitar(`/documentos-rh/${encodeURIComponent(idDocumento)}/conteudo`, {
    method: 'GET',
  });
}

export async function baixarDocumentoRh(idDocumento) {
  return requisitarArquivo(
    `/documentos-rh/${encodeURIComponent(idDocumento)}/download`,
    { method: 'GET' },
  );
}

export async function renomearDocumentoRh(idDocumento, nome) {
  return requisitar(`/documentos-rh/${encodeURIComponent(idDocumento)}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ nome }),
  });
}

export async function excluirDocumentoRh(idDocumento) {
  return requisitar(`/documentos-rh/${encodeURIComponent(idDocumento)}`, {
    method: 'DELETE',
  });
}

File: apps/backend/rh_api/routers/documentos_rh.py
from __future__ import annotations

from fastapi import APIRouter, Depends, File, Query, UploadFile
from fastapi.responses import FileResponse

from ..auth import AuthenticatedUser
from ..dependencies import audit_action, get_current_user, get_repository, require_permissions
from ..repositories import DatabaseRepository
from ..schemas.documentos_rh import (
    CriarPastaDocumentoRhRequest,
    RenomearDocumentoRhRequest,
)


router = APIRouter(tags=["documentos-rh"], dependencies=[Depends(get_current_user)])


@router.get(
    "/documentos-rh",
    dependencies=[Depends(require_permissions("documentos_rh.visualizar"))],
)
def listar_documentos_rh(
    id_pasta_pai: int | None = Query(default=None),
    tipo: str = Query(default=""),
    categoria_extensao: str = Query(default=""),
    busca: str = Query(default=""),
    criado_por: str = Query(default=""),
    data_criacao_de: str = Query(default=""),
    data_criacao_ate: str = Query(default=""),
    data_modificacao_de: str = Query(default=""),
    data_modificacao_ate: str = Query(default=""),
    repository: DatabaseRepository = Depends(get_repository),
):
    itens = repository.list_documentos_rh(
        id_pasta_pai=id_pasta_pai,
        tipo=tipo,
        categoria_extensao=categoria_extensao,
        busca=busca,
        criado_por=criado_por,
        data_criacao_de=data_criacao_de,
        data_criacao_ate=data_criacao_ate,
        data_modificacao_de=data_modificacao_de,
        data_modificacao_ate=data_modificacao_ate,
    )
    trilha = repository.get_documento_rh_breadcrumb(id_pasta_pai)
    return {"itens": itens, "trilha": trilha}


@router.post(
    "/documentos-rh/pastas",
    dependencies=[Depends(require_permissions("documentos_rh.gerenciar"))],
)
def criar_pasta_documento_rh(
    payload: CriarPastaDocumentoRhRequest,
    user: AuthenticatedUser = Depends(get_current_user),
    repository: DatabaseRepository = Depends(get_repository),
):
    resultado = repository.criar_pasta_documento_rh(
        nome=payload.nome,
        id_pasta_pai=payload.id_pasta_pai,
        criado_por=user.nome or user.username,
    )
    audit_action(
        repository,
        user,
        modulo="Drive-Conecta",
        acao="criar_pasta",
        entidade="documento_rh",
        entidade_id=str(resultado.get("id", "")),
        valor_novo={"nome": payload.nome, "id_pasta_pai": payload.id_pasta_pai},
    )
    return resultado


@router.post(
    "/documentos-rh/upload",
    dependencies=[Depends(require_permissions("documentos_rh.gerenciar"))],
)
async def upload_documento_rh(
    arquivo: UploadFile = File(...),
    id_pasta_pai: int | None = Query(default=None),
    user: AuthenticatedUser = Depends(get_current_user),
    repository: DatabaseRepository = Depends(get_repository),
):
    resultado = await repository.upload_arquivo_documento_rh(
        arquivo=arquivo,
        id_pasta_pai=id_pasta_pai,
        criado_por=user.nome or user.username,
    )
    audit_action(
        repository,
        user,
        modulo="Drive-Conecta",
        acao="upload_arquivo",
        entidade="documento_rh",
        entidade_id=str(resultado.get("id", "")),
        valor_novo={"nome": arquivo.filename or "", "id_pasta_pai": id_pasta_pai},
    )
    return resultado


@router.get(
    "/documentos-rh/{id_documento}/conteudo",
    dependencies=[Depends(require_permissions("documentos_rh.visualizar"))],
)
def obter_conteudo_documento_rh(
    id_documento: int,
    repository: DatabaseRepository = Depends(get_repository),
):
    return repository.obter_conteudo_documento_rh(id_documento)


@router.get(
    "/documentos-rh/{id_documento}/download",
    dependencies=[Depends(require_permissions("documentos_rh.visualizar"))],
)
def baixar_documento_rh(
    id_documento: int,
    repository: DatabaseRepository = Depends(get_repository),
):
    arquivo = repository.baixar_documento_rh(id_documento)
    return FileResponse(
        arquivo["path"],
        media_type=arquivo["media_type"],
        filename=arquivo["filename"],
    )


@router.patch(
    "/documentos-rh/{id_documento}",
    dependencies=[Depends(require_permissions("documentos_rh.gerenciar"))],
)
def renomear_documento_rh(
    id_documento: int,
    payload: RenomearDocumentoRhRequest,
    user: AuthenticatedUser = Depends(get_current_user),
    repository: DatabaseRepository = Depends(get_repository),
):
    resultado = repository.renomear_documento_rh(id_documento, payload.nome)
    audit_action(
        repository,
        user,
        modulo="Drive-Conecta",
        acao="renomear_item",
        entidade="documento_rh",
        entidade_id=str(id_documento),
        valor_novo={"nome": payload.nome},
    )
    return resultado


@router.delete(
    "/documentos-rh/{id_documento}",
    dependencies=[Depends(require_permissions("documentos_rh.gerenciar"))],
)
def excluir_documento_rh(
    id_documento: int,
    user: AuthenticatedUser = Depends(get_current_user),
    repository: DatabaseRepository = Depends(get_repository),
):
    resultado = repository.excluir_documento_rh(id_documento)
    audit_action(
        repository,
        user,
        modulo="Drive-Conecta",
        acao="excluir_item",
        entidade="documento_rh",
        entidade_id=str(id_documento),
    )
    return resultado
